### PR TITLE
app-sso csharp

### PR DESCRIPTION
**Pull request by Swapna-MSFT**
# App SSO C#

This app talks about the Teams Tab, Bot, ME - search, action, linkunfurl SSO with C#

__Tab SSO__
This sample shows how to implement Azure AD single sign-on support for tabs. It will

- Obtain an access token for the logged-in user using SSO
- Call a web service - also part of this project - to exchange this access token
- Call Graph and retrieve the user's profile

__Bot, ME SSO__
Bot Framework v4 bot using Teams authentication

This bot has been created using [Bot Framework](https://dev.botframework.com), it shows how to get started with authentication in a bot for Microsoft Teams.

The focus of this sample is how to use the Bot Framework support for oauth in your bot. Teams behaves slightly differently than other channels in this regard. Specifically an Invoke Activity is sent to the bot rather than the Event Activity used by other channels. _This Invoke Activity must be forwarded to the dialog if the OAuthPrompt is being used._ This is done by subclassing the ActivityHandler and this sample includes a reusable TeamsActivityHandler. This class is a candidate for future inclusion in the Bot Framework SDK.

### CODE REVIEW

File: samples/app-sso/csharp/app_sso/Bots/DialogBot.cs [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+                                        Title = "Bot Service OAuth",
+                                    },
+                                },
+                            },
+                        },
+                    };
+                }
+
+                var client = new SimpleGraphClient(tokenResponse.Token);
+
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/Bots/DialogBot.cs:257]: Fix spacing here

File: samples/app-sso/csharp/app_sso/Bots/DialogBot.cs [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+                    },
+                };
+            }
+            return null;
+        }
+
+        private async Task<TokenResponse> GetTokenResponse(ITurnContext<IInvokeActivity> turnContext, string state, CancellationToken cancellationToken)
+        {
+            var magicCode = string.Empty;
+
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/Bots/DialogBot.cs:309]: Have one common method for magic code and tokenResponse like this: [https://github.com/microsoft/BotBuilder-Samples/blob/f03627c0662927702c519c9e05365d0d74aa8357/samples/csharp_dotnetcore/52.teams-messaging-extensions-search-auth-config/Bots/TeamsMessagingExtensionsSearchAuthConfigBot.cs#L48](https://github.com/microsoft/BotBuilder-Samples/blob/f03627c0662927702c519c9e05365d0d74aa8357/samples/csharp_dotnetcore/52.teams-messaging-extensions-search-auth-config/Bots/TeamsMessagingExtensionsSearchAuthConfigBot.cs#L48)

File: samples/app-sso/csharp/app_sso/Bots/DialogBot.cs [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+                 turnContext,
+                 _connectionName,
+                 turnContext.Activity.From.Id,
+                 new TokenExchangeRequest
+                 {
+                     Token = tokenExchangeRequest.Token,
+                 },
+                 cancellationToken).ConfigureAwait(false);
+            }
+#pragma warning disable CA1031 //Do not catch general exception types (ignoring, see comment below)
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/Bots/DialogBot.cs:363]: is this compile perfectly?

File: samples/app-sso/csharp/app_sso/Bots/TeamsMessagingExtensionsSearchAuthConfigBot.cs [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+            // When the user submits the settings page, this event is fired.
+            if (settings["state"] != null)
+            {
+                var userConfigSettings = settings["state"].ToString();
+                await _userConfigProperty.SetAsync(turnContext, userConfigSettings, cancellationToken);
+            }
+        }
+
+        protected override async Task<MessagingExtensionResponse> OnTeamsMessagingExtensionQueryAsync(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionQuery action, CancellationToken cancellationToken)
+        {
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/Bots/TeamsMessagingExtensionsSearchAuthConfigBot.cs:132]: it seems this file is not required, could you please verify once? most of the methods are implemented in DialogBot.cs file, please remove this file if not required

File: samples/app-sso/csharp/app_sso/Helper/SSOAuthHelper.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        /// </summary>
+        /// <param name="configuration">IConfiguration instance.</param>
+        /// <returns>Valid Issuers.</returns>
+        public static IEnumerable<string> GetValidIssuers(IConfiguration configuration)
+        {
+            var tenantId = configuration[TenantIdConfigurationSettingsKey];
+
+            var validIssuers = GetSettings(configuration);
+
+            validIssuers = validIssuers.Select(validIssuer => validIssuer.Replace("TENANT_ID", tenantId));
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/Helper/SSOAuthHelper.cs:80]: Please mention in the ReadME to update this tenantId

File: samples/app-sso/csharp/app_sso/Resources/justintimeinstallation.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
+{
+  "type": "AdaptiveCard",
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/Resources/justintimeinstallation.json:2]: We don't need justinTime installation card here please remove this, Let me know if it is being used

File: samples/app-sso/csharp/app_sso/SimpleGraphClient.cs [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+            {
+                throw new ArgumentNullException(nameof(token));
+            }
+
+            _token = token;
+        }
+
+        // Sends an email on the users behalf using the Microsoft Graph API
+        public async Task SendMailAsync(string toAddress, string subject, string content)
+        {
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/SimpleGraphClient.cs:33]: SendMailAsync method is not being used anywhere in this sample app, Please cleanup the unused code

File: samples/app-sso/csharp/app_sso/SimpleGraphClient.cs [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+                ToRecipients = recipients,
+            };
+
+            // Send the message.
+            await graphClient.Me.SendMail(email, true).Request().PostAsync();
+        }
+
+        // Gets mail for the user using the Microsoft Graph API
+        public async Task<Message[]> GetRecentMailAsync()
+        {
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/SimpleGraphClient.cs:79]: Please clean up unused code

File: samples/app-sso/csharp/app_sso/SimpleGraphClient.cs [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+            {
+                return null;
+            }
+
+            return uploadedFile;
+        }
+
+        public  async Task<Permission> GetSharingLinkAsync(string Id)
+        {
+            Permission permission = null;
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/SimpleGraphClient.cs:179]: is this method used anywhere?

File: samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json [hunk 10]
```diff
@@ -0,0 +1,136 @@
+{
+  "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.7/MicrosoftTeams.schema.json",
+  "manifestVersion": "1.7",
+  "version": "1.0.0",
+  "id": "<<YOUR-MICROSOFT-APP-ID>>",
+  "packageName": "com.microsoft.teams.samples",
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json:6]: rename the package name to `com.microsoft.teams.samples.AppSSO`

File: samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    "websiteUrl": "https://example.azurewebsites.net",
+    "privacyUrl": "https://example.azurewebsites.net/privacy",
+    "termsOfUseUrl": "https://example.azurewebsites.net/termsofuse"
+  },
+  "icons": {
+    "color": "icon-color.png",
+    "outline": "icon-outline.png"
+  },
+  "name": {
+    "short": "BotTabMeSSO",
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json:18]: Rename this sample app name to `App SSO Node JS`

File: samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    "full": "Teams Single Sign On Sample"
+  },
+  "description": {
+    "short": "Bot Tab ME SSO Scenario",
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json:22]: Update description to `This sample app demonstrates SSO for Tab Bot and ME (Link Unfurl, Action and Search Config)`

File: samples/app-sso/csharp/app_sso/appsettings.json [hunk 13]
```diff
@@ -0,0 +1,16 @@
+{
+  "MicrosoftAppId": "<<YOUR-MICROSOFT-APP-ID>>",
+  "MicrosoftAppPassword": "<<YOUR-MICROSOFT-APP-PASSWORD>>",
+  "ConnectionName": "<<YOUR-CONNECTION-NAME>>",
+  "SiteUrl": "<<DOMAIN-NAME>>",
+  "AzureAd": {
+    "Instance": "https://login.microsoftonline.com/",
+    "TenantId": "<<YOUR-MICROSOFT-APP-TENANT-ID>>",
+    "ClientId": "<<YOUR-MICROSOFT-APP-ID>>",
```
Comment (Trinetra-MSFT) [inline samples/app-sso/csharp/app_sso/appsettings.json:9]: There is no need to specify two time client ID and secret, maintain it in one config key and use it everywhere

File: samples/app-sso/csharp/README.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+The sample uses the bot authentication capabilities in [Azure Bot Service](https://docs.botframework.com), providing features to make it easier to develop a bot that authenticates users to various identity providers such as Azure AD (Azure Active Directory), GitHub, Uber, etc. The OAuth token is then used to make basic Microsoft Graph queries. Refer the **SSO** setup [documentation](https://docs.microsoft.com/en-us/microsoftteams/platform/bots/how-to/authentication/add-authentication?tabs=dotnet%2Cdotnet-sample).
+
+> IMPORTANT: The manifest file in this app adds "token.botframework.com" to the list of `validDomains`. This must be included in any bot that uses the Bot Framework OAuth flow.
+
+## Prerequisites
+
```
[the remark this answers is not in the document]
Comment (Swapna-MSFT) [reply samples/app-sso/csharp/README.md:24]: We already added the give a link to set a connection string

File: samples/app-sso/csharp/app_sso/Bots/DialogBot.cs [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+        {
+            try
+            {
+                await base.OnTurnAsync(turnContext, cancellationToken);
+                // After the turn is complete, persist any UserState changes.
+                await _userState.SaveChangesAsync(turnContext);
+            }
+            catch (Exception ex)
+            {
+            }
```
[the remark this answers is not in the document]
Comment (Swapna-MSFT) [reply samples/app-sso/csharp/app_sso/Bots/DialogBot.cs:58]: Console.write added to log the excetion

File: samples/app-sso/csharp/app_sso/Bots/DialogBot.cs [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+                };
+            }
+
+            var client = new SimpleGraphClient(tokenResponse.Token);
+            var profile = await client.GetMyProfile();
+            var imagelink = await client.GetPhotoAsync();
+            var heroCard = new ThumbnailCard
+            {
+                Title = "Thumbnail Card",
+                Text = $"Hello, {profile.DisplayName}",
```
[the remark this answers is not in the document]
Comment (Swapna-MSFT) [reply samples/app-sso/csharp/app_sso/Bots/DialogBot.cs]: Changes updated

File: samples/app-sso/csharp/app_sso/Bots/DialogBot.cs [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+                    {
+                        Type = "result",
+                        AttachmentLayout = "list",
+                        Attachments = new List<MessagingExtensionAttachment> { attachment }
+                    }
+                };
+            }
+            catch (Exception ex)
+            {
+
```
[the remark this answers is not in the document]
Comment (Swapna-MSFT) [reply samples/app-sso/csharp/app_sso/Bots/DialogBot.cs]: Resolved the spaces issues

File: samples/app-sso/csharp/app_sso/Bots/DialogBot.cs (hunk 1 above)
Comment (Swapna-MSFT) [reply samples/app-sso/csharp/app_sso/Bots/DialogBot.cs]: Fixed the Spaces issues

File: samples/app-sso/csharp/app_sso/SimpleGraphClient.cs (hunk 9 above)
Comment (Swapna-MSFT) [reply samples/app-sso/csharp/app_sso/SimpleGraphClient.cs]: Removed not used functions

File: samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json (hunk 12 above)
Comment (Swapna-MSFT) [reply samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json]: Description is  updated on Manifest

File: samples/app-sso/csharp/app_sso/appsettings.json (hunk 13 above)
Comment (Swapna-MSFT) [reply samples/app-sso/csharp/app_sso/appsettings.json:9]: image

To get Azured id options we need to add these fields are mandatory

File: samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json (hunk 11 above)
Comment (Swapna-MSFT) [reply samples/app-sso/csharp/app_sso/TeamsAppManifest/manifest.json]: Renamed as App sso csharp

**Review by Trinetra-MSFT — APPROVED** [review]

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**